packages/web-components/fast-foundation/src/virtual-list/virtual-list.ts

```diff
@@ -0,0 +1,885 @@
+                newFirstRenderedIndex = this.spanmap.indexOf(firstVisibleItem);
+                const lastVisibleItem: SpanMap | undefined = this.spanmap.find(
+                    x => x.start >= renderedRangeEnd
+                );
+                newLastRenderedIndex =
+                    lastVisibleItem === undefined
+                        ? this.spanmap.length - 1
+                        : this.spanmap.indexOf(lastVisibleItem);
+            }
+
+            newFirstRenderedIndex = Math.max(0, newFirstRenderedIndex);
+            newLastRenderedIndex = Math.min(newLastRenderedIndex, this.items.length - 1);
+
+            this.startSpacerSpan = this.spanmap[newFirstRenderedIndex].start;
+            this.endSpacerSpan =
+                this.spanmap[this.spanmap.length - 1].end -
+                this.spanmap[newLastRenderedIndex].end;
+        }
+
+        const newVisibleItems = this.items.slice(
+            newFirstRenderedIndex,
+            newLastRenderedIndex + 1
+        );
+
+        this.updateVisibleItemSpans(newFirstRenderedIndex, newLastRenderedIndex);
+        this.visibleItems.splice(0, this.visibleItems.length, ...newVisibleItems);
+    }
+
+    /**
+     *  Updates the span map
+     */
+    private updateVisibleItemSpans(
+        newFirstRenderedIndex: number,
+        newLastRenderedIndex: number
+    ): void {
+        const newVisibleItemSpans: SpanMap[] = [];
+
+        let top: number = this.startSpacerSpan;
+
+        if (this.spanmap === undefined) {
+            for (let i: number = newFirstRenderedIndex; i <= newLastRenderedIndex; i++) {
+                const thisSpanMap: SpanMap = {
+                    start: top,
+                    end: top + this.itemSpan,
+                    span: this.itemSpan,
+                };
+                top = thisSpanMap.end;
+                newVisibleItemSpans.push(thisSpanMap);
+            }
+
+            this.visibleItemSpans.splice(
+                0,
+                this.visibleItemSpans.length,
+                ...newVisibleItemSpans
+            );
+        } else {
+            this.visibleItemSpans.splice(
+                0,
+                this.visibleItemSpans.length,
+                ...this.spanmap.slice(newFirstRenderedIndex, newLastRenderedIndex + 1)
+            );
+        }
+
+        this.updateRenderedRange(newFirstRenderedIndex, newLastRenderedIndex);
+    }
+
+    /**
+     *  Updates the range of rendered items
+     */
+    private updateRenderedRange(
+        newFirstRenderedIndex: number,
+        newLastRenderedIndex: number
+    ): void {
+        if (
+            newFirstRenderedIndex === this.firstRenderedIndex &&
+            newLastRenderedIndex === this.lastRenderedIndex
+        ) {
+            return;
+        }
+
+        this.firstRenderedIndex = newFirstRenderedIndex;
+        this.lastRenderedIndex = newLastRenderedIndex;
+
+        this.$emit("rendered-range-change", this, { bubbles: false });
```

Comment:
We're not consistent on this, the doc just says "lower-case", but should we have the hyphens? Standard html elements don't.

Reply:
Apparently, I'm guilty of doing both "alllowercasewithwordsmungedtogether" and "all-lower-case-with-hyphens".  What should be the standard?

Reply:
no hyphens it is